### PR TITLE
rpi_pico: Add hwinfo support

### PR DESCRIPTION
**Pull request by yonsch**
Adds hwinfo support for the rpi_pico series. Additionally, support for calling bootrom functions was added.

### CODE REVIEW

**Review by tejlmand — CHANGES_REQUESTED** [review]
Generally looks good.

Minor request regarding compiler flag.

File: modules/hal_rpi_pico/CMakeLists.txt [hunk 1]
```diff
@@ -80,4 +80,21 @@ if(CONFIG_HAS_RPI_PICO)
...
+  # Originally pico-sdk places them in the RW data section, so this
+  # implementation does the same.
+  zephyr_linker_sources_ifdef(CONFIG_PICOSDK_USE_FLASH RWDATA timecritical.ld)
+
+  # A function in flash.c adds 1 to a function pointer, causing a warning
+  set_source_files_properties(
+    ${rp2_common_dir}/hardware_flash/flash.c
+    PROPERTIES
+    COMPILE_FLAGS "-Wno-pointer-arith"
+  )
```
[the remark this answers is not in the document]
Comment (yonsch) [reply modules/hal_rpi_pico/CMakeLists.txt]: @tejlmand done

File: drivers/hwinfo/hwinfo_rpi_pico.c [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+	*cause = flags;
+	return 0;
+}
+
+int z_impl_hwinfo_clear_reset_cause(void)
+{
+	/* The reset register can't be erased, nothing to do. */
+
+	return 0;
```
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_rpi_pico.c:57]: -ENOSUP would be more precise here.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

**Comment by alexanderwachter** [comment]
@yonsch I have to apologize. I told you the wrong errno.
-ENOSYS is the correct one. The API doc is not up to date.

**Review by tejlmand — DISMISSED** [review]
lgtm

**Review by alexanderwachter — DISMISSED** [review]
Thanks for the doc change!

**Review by tejlmand — APPROVED** [review]
lgtm - re-approved

**Review by de-nordic — APPROVED** [review]

**Review by alexanderwachter — APPROVED** [review]

**Pull request merged**